dynamodb/bucket.go

```diff
@@ -0,0 +1,158 @@
+}
+
+// Remaining space in the bucket.
+func (b *bucket) Remaining() uint {
+	return b.remaining
+}
+
+// Reset returns when the bucket will be drained.
+func (b *bucket) Reset() time.Time {
+	return b.reset
+}
+
+// Add to the bucket. This isn't a traditional leakybucket. It does not "drip" at a steady rate.
+// Rather, it treats the rate as a window for the specified capacity. If amount exceeds the
+// remaining capacity at any stage during the sliding window it will return `leakybucket.ErrorFull`
+// as expected
+func (b *bucket) Add(amount uint) (leakybucket.BucketState, error) {
+	// db.GetItem(bucket)
+	// if expired -> flush
+	// if (value + incrementAmount > capacity) -> error
+	// increment by amount -> return error if any
+	dbBucket, err := b.db.bucket(b.name)
+	if err != nil {
+		return b.state(), err
+	}
+	if dbBucket.isExpired() {
+		dbBucket, err = b.db.flushBucket(*dbBucket, time.Now().Add(b.rate))
+		if err != nil {
+			if err != errConcurrentFlush {
+				return b.state(), err
+			}
+			// if another consumer flushed the bucket re-fetch
+			dbBucket, err = b.db.bucket(b.name)
+			if err != nil {
+				return b.state(), err
+			}
+		}
+	}
+	// update local state
+	b.remaining = b.capacity - min(dbBucket.Value, b.capacity)
+	b.reset = dbBucket.Expiration
+	if amount > b.remaining {
+		return b.state(), leakybucket.ErrorFull
+	}
+	updatedDBBucket, err := b.db.incrementBucketValue(b.name, amount, b.capacity)
+	if err != nil {
+		if err == errBucketCapacityExceeded {
+			return b.state(), leakybucket.ErrorFull
+		}
+		return b.state(), err
+	}
+	// ensure we can't overflow
+	b.remaining = b.capacity - min(uint(updatedDBBucket.Value), b.capacity)
+	return b.state(), nil
+}
+
+func (b *bucket) state() leakybucket.BucketState {
+	return leakybucket.BucketState{
+		Capacity:  b.Capacity(),
+		Remaining: b.Remaining(),
+		Reset:     b.Reset(),
+	}
+}
+
+var _ leakybucket.Storage = &Storage{}
+
+// Storage is a dyanamodb-based, non thread-safe leaky bucket factory.
+type Storage struct {
+	db db
+}
+
+// Create a bucket. It will determine the current state of the bucket based on:
+// - The corresponding bucket in the database
+// - From scratch using the values provided
+func (s *Storage) Create(name string, capacity uint, rate time.Duration) (leakybucket.Bucket, error) {
+	bucket := &bucket{
+		name:      dbKey(name),
+		capacity:  capacity,
+		remaining: capacity,
+		reset:     time.Now().Add(rate),
+		rate:      rate,
+		db:        s.db,
+	}
+	dbBucket, err := s.db.bucket(bucket.name)
+	if err != nil {
+		// bubble up all errors other than a non-existing bucket
+		if err != errBucketNotFound {
+			return nil, err
+		}
+		if err := s.db.createBucket(newDDBBucket(name, bucket.reset)); err != nil {
+			return nil, err
+		}
+	} else {
+		// guarantee the bucket is in a good state
+		if dbBucket.isExpired() {
+			if _, err := bucket.Add(0); err != nil {
+				return nil, err
+			}
+		}
+		bucket.remaining = capacity - dbBucket.Value
```

Comment:
is there a clearer name than `Value`? It seems to be the complement of capacity but it's not obvious from the naming. I can't think of a better name though. `Water` would seem to be the name that follows the leaky bucket theme but it seems a little meh. `Added` might denote that the value is added to

Reply:
also shouldn't this be `max(capacity-dbBucket.Value, 0)`?

Reply:
I was having the same problem with naming here. I went with value because I'm thinking of this as a key value store with some extra house keeping. I'll make sure to iterate on the name a bit prior to merging

>also shouldn't this be max(capacity-dbBucket.Value, 0)?

👍

Reply:
FWIW when I read `Value` I did not see the connotation around the way in which the database is getting used as a key-value store. Even if I did I think that connotation is somewhat nebulous. What does it mean to use a database as a key-value store, and how does it help me understand the code?